File: DDD DSL/DDD DSL Design War Measurement System/Code/Java/src/fayelab/ddd/measure/vecarith/VecOp.java
package fayelab.ddd.measure.vecarith;

import java.util.ArrayList;
import java.util.Collections;
import java.util.List;
import java.util.stream.Collectors;
import java.util.stream.IntStream;

public class VecOp
{
    public static boolean equal(List<Integer> vec1, List<Integer> vec2)
    {
        return vec1.equals(vec2);
    }

    public static List<Integer> add(List<Integer> vec1, List<Integer> vec2)
    {
        return IntStream.range(0, vec1.size())
                        .mapToObj(i -> vec1.get(i) + vec2.get(i))
                        .collect(Collectors.toList());
    }

    public static int dotProduct(List<Integer> rowVec, List<Integer> colVec)
    {
        return IntStream.range(0, rowVec.size())
                        .mapToObj(i -> rowVec.get(i) * colVec.get(i))
                        .reduce(0, (x, y) -> x + y);
    }

    public static List<Integer> div(List<Integer> dividendVec, List<Integer> divisorVec)
    {
        List<Integer> rDividendVec = reverse(dividendVec);
        List<Integer> rDivisorVec = reverse(divisorVec);
        int[] quotient = new int[]{0};

        List<Integer> rResultVec = IntStream.range(0, rDividendVec.size())
                                            .collect(() -> new ArrayList<>(),
                                                     (acc, i) -> {
                                                         int dividend = rDividendVec.get(i) + quotient[0];
                                                         int divisor = rDivisorVec.get(i);
                                                         quotient[0] = dividend / divisor;
                                                         acc.add(dividend % divisor);
                                                     },
                                                     (acc1, acc2) -> acc1.addAll(acc2));
        return reverse(rResultVec);
    }

    public static List<Integer> dotDiv(int baseValue, List<Integer> vec)
    {
        int[] remainder = new int[]{baseValue};

        return vec.stream()
                  .collect(() -> new ArrayList<>(),
                           (acc, c) -> {
                               int quotient = remainder[0] / c;
                               remainder[0] = remainder[0] % c;
                               acc.add(quotient);
                           },
                           (acc1, acc2) -> acc1.addAll(acc2));
    }

    public static List<Integer> scale(int c, List<Integer> vec)
    {
        return vec.stream().map(x -> c * x).collect(Collectors.toList());
    }

    private static List<Integer> reverse(List<Integer> vec)
    {
        List<Integer> rVec = new ArrayList<>(vec);
        Collections.reverse(rVec);
        return rVec;
    }
}

File: DDD DSL/DDD DSL Design War Measurement System/Code/Python/vector/scale/vecop.py
def equal(v1, v2):
    return v1 == v2

def add(v1, v2):
    return [vc1 + vc2 for (vc1, vc2) in zip(v1, v2)]

def dotproduct(v1, v2):
    return sum([vc1 * vc2 for (vc1, vc2) in zip(v1, v2)])

def div(v1, v2):
    result = []
    quotient = 0
    for (vc1, vc2) in zip(v1[::-1], v2[::-1]):
        dividend = vc1 + quotient
        divisor = vc2
        quotient = dividend / divisor
        result.append(dividend % divisor)
    return result[::-1]

def div2(v1, v2):
    def acc_div2((result, wquotient), (vc1, vc2)):
        dividend = vc1 + wquotient[0]
        divisor = vc2
        wquotient[0] = dividend / divisor
        result.append(dividend % divisor)
        return result, wquotient

    return reduce(acc_div2, zip(v1[::-1], v2[::-1]), ([], [0]))[0][::-1]

def dotdiv(c, v):
    result = []
    remainder = c
    for vc in v:
        quotient = remainder / vc
        remainder = remainder % vc
        result.append(quotient)
    return result

def dotdiv2(c, v):
    def acc_dotdiv2((v, wc), vc):
        quotient = wc[0] / vc
        wc[0] = wc[0] % vc
        v.append(quotient)
        return v, wc

    return reduce(acc_dotdiv2, v, [[], [c]])[0]

def scale(c, v):
    return [c * vc for vc in v]